src/front/wgsl/lexer.rs

```diff
@@ -13,48 +13,306 @@ fn consume_any(input: &str, what: impl Fn(char) -> bool) -> (&str, &str) {
+                ..
+            } => hex || leading_zeros == 1, // No leading zeros allowed in non-hex integers, "0" is always allowed.
+            Self {
+                _minus: minus,
+                hex,
+                leading_zeros,
+                digit_state: NLDigitState::DigitBeforeDot,
+                uint_suffix,
+            } => {
+                (hex || leading_zeros == 0) // No leading zeros allowed in non-hex integers.
+                                              // In this state the number has non-zero digits,
+                                              // i.e. it is not just "0".
+                    && (minus ^ uint_suffix) // Either a negative number, or and unsigned integer, not both.
+            }
+            _ => self.is_float(),
+        }
+    }
+
+    pub fn is_float(&self) -> bool {
+        !self.uint_suffix
+            && (self.digit_state == NLDigitState::DigitsThenDot
+                || self.digit_state == NLDigitState::DigitAfterDot
+                || self.digit_state == NLDigitState::DigitAfterExponent)
+    }
+}
+
 fn consume_number(input: &str) -> (Token, &str) {
-    //Note: I wish this function was simpler and faster...
-    let mut is_first_char = true;
-    let mut right_after_exponent = false;
+    let (minus, working_substr, minus_offset) = try_skip_prefix(input, "-");
+
+    let (hex, working_substr, hex_offset) = try_skip_prefix(working_substr, "0x");
+
+    let mut state = NumberLexerState {
+        _minus: minus,
+        hex,
+        leading_zeros: 0,
+        digit_state: NLDigitState::Nothing,
+        uint_suffix: false,
+    };
 
     let mut what = |c| {
-        if is_first_char {
-            is_first_char = false;
-            c == '-' || ('0'..='9').contains(&c) || c == '.'
-        } else if c == 'e' || c == 'E' {
-            right_after_exponent = true;
-            true
-        } else if right_after_exponent {
-            right_after_exponent = false;
-            ('0'..='9').contains(&c) || c == '-'
-        } else {
-            ('0'..='9').contains(&c) || c == '.'
+        match state {
+            NumberLexerState {
+                hex,
+                digit_state: NLDigitState::Nothing,
+                uint_suffix: false,
+                ..
+            } => match c {
+                '0' => {
+                    state.digit_state = NLDigitState::LeadingZero;
+                    state.leading_zeros += 1;
+                    true
+                }
+                '1'..='9' => {
+                    state.digit_state = NLDigitState::DigitBeforeDot;
+                    true
+                }
+                'a'..='f' | 'A'..='F' if hex => {
+                    state.digit_state = NLDigitState::DigitBeforeDot;
+                    true
+                }
+                '.' => {
+                    state.digit_state = NLDigitState::OnlyDot;
+                    true
+                }
+                _ => false,
```

Comment:
nit: would it be easier to just change this to `return false` to avoid adding `true` to each and every handled case?